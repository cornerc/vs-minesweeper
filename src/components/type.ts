export type BaseToggles = "drawer" | "configDialog" | "infoDialog";
export type SingleToggles = "scoreAlert" | "snackbar";
export type TurnToggles = "waitingDialog";

export interface SideMenuItems {
  icon: string;
  title: string;
  text: string;
  to: object;
}

export interface HeaderCenterItems {
  icon: string;
  title: string;
  click: Function;
  content: Function;
}

export interface HeaderRightItems {
  icon: string;
  title: string;
  click: Function;
}

export interface Config {
  darkTheme: boolean;
  mine: number;
  row: number;
  col: number;
}
