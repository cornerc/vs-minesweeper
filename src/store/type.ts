export interface Field {
  mine: number;
  row: number;
  col: number;
  cells: Cell[];
}

export interface Cell {
  isOpen: boolean;
  isFlag: boolean;
  isLandMine: boolean;
  aroundMines: number;
}
