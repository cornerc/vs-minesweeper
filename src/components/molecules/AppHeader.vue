<template>
  <v-toolbar dense flat :tile="false">
    <v-toolbar-title title="vs-minesweeper">
      VSマインスイーパー
    </v-toolbar-title>
    <v-spacer />
    <template v-for="item in headerCenterItems">
      <span :key="item.title">
        <v-icon>{{ item.icon }}</v-icon>
        <v-chip class="ma-2" :title="item.title" label @click.stop="item.click">
          {{ item.content() }}
        </v-chip>
      </span>
      <v-spacer :key="item.title + '-spacer'" />
    </template>
    <template v-for="item in headerRightItems">
      <v-btn
        :key="item.title"
        :title="item.title"
        icon
        @click.stop="item.click"
      >
        <v-icon>{{ item.icon }}</v-icon>
      </v-btn>
    </template>
  </v-toolbar>
</template>

<script lang="ts">
import {Component, Prop, Vue} from "vue-property-decorator";
import {HeaderCenterItems, HeaderRightItems} from "@/components/type";

@Component
export default class AppHeader extends Vue {
  @Prop({type: Array, default: () => []})
  private headerCenterItems: HeaderCenterItems[];
  @Prop({type: Array, default: () => []})
  private headerRightItems: HeaderRightItems[];
}
</script>
