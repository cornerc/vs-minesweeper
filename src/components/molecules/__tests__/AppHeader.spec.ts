import "@/tests/setup.js";
import Vuetify from "vuetify";
import {mount, createLocalVue, Wrapper} from "@vue/test-utils";
import AppHeader from "@/components/molecules/AppHeader.vue";
import {HeaderCenterItems, HeaderRightItems} from "@/components/type";

const localVue = createLocalVue();

describe("AppHeader", () => {
  let vuetify: any;
  let mountWrapper: (options?: object) => Wrapper<Vue>;
  beforeEach(() => {
    vuetify = new Vuetify();
    mountWrapper = (options = {}) => {
      return mount(AppHeader, {
        localVue,
        vuetify,
        ...options,
      });
    };
  });

  it("should be have title by default", () => {
    const wrapper = mountWrapper({
      propsData: {},
    });
    const title = wrapper.find(".v-toolbar__title");
    expect(title.text()).toBe("VSマインスイーパー");

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should be exist center item", () => {
    const headerCenterItems: HeaderCenterItems[] = [
      {
        icon: "mdi-table",
        title: "table",
        content: "table-content",
      },
      {
        icon: "mdi-av-timer",
        title: "time",
        content: "time-content",
      },
    ];
    const wrapper = mountWrapper({
      propsData: {
        headerCenterItems,
      },
    });

    for (let i = 0; i < headerCenterItems.length; i++) {
      const item = headerCenterItems[i];

      const icon = wrapper.find("i." + item.icon);
      expect(icon.exists()).toBe(true);

      const content = wrapper.find(`[title=${item.title}]`);
      expect(content.exists()).toBe(true);
      expect(content.text()).toBe(item.content);

      content.trigger("click");

      const emittedRes = wrapper.emitted();

      expect(emittedRes.clickCenterContent).toBeDefined();
      if (emittedRes.clickCenterContent) {
        expect(emittedRes.clickCenterContent[i][0]).toEqual(item.title);
      }
    }

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should be exist right item", () => {
    const headerRightItems: HeaderRightItems[] = [
      {
        icon: "mdi-information",
        title: "information",
      },
    ];
    const wrapper = mountWrapper({
      propsData: {
        headerRightItems,
      },
    });
    for (let i = 0; i < headerRightItems.length; i++) {
      const item = headerRightItems[i];
      const icon = wrapper.find("i." + item.icon);
      expect(icon.exists()).toBe(true);

      const button = wrapper.find(`[title=${item.title}]`);

      button.trigger("click");

      const emittedRes = wrapper.emitted();

      expect(emittedRes.clickRightContent).toBeDefined();
      if (emittedRes.clickRightContent) {
        expect(emittedRes.clickRightContent[i][0]).toEqual(item.title);
      }
    }

    expect(wrapper.html()).toMatchSnapshot();
  });
});
