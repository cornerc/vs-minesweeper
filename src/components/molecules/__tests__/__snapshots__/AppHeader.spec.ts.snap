// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppHeader should be exist center item 1`] = `
<header class="v-sheet v-sheet--tile theme--light v-toolbar v-toolbar--dense v-toolbar--flat primary" style="height: 48px;">
  <div class="v-toolbar__content" style="height: 48px;">
    <div title="vs-minesweeper" class="v-toolbar__title">
      VSマインスイーパー
    </div>
    <div class="spacer"></div> <span><i aria-hidden="true" class="v-icon notranslate mdi mdi-table theme--light"></i> <span class="ma-2 v-chip v-chip--clickable v-chip--label v-chip--no-color theme--light v-size--default" title="table"><span class="v-chip__content">
        table-content
      </span></span></span>
    <div class="spacer"></div><span><i aria-hidden="true" class="v-icon notranslate mdi mdi-av-timer theme--light"></i> <span class="ma-2 v-chip v-chip--clickable v-chip--label v-chip--no-color theme--light v-size--default" title="time"><span class="v-chip__content">
        time-content
      </span></span></span>
    <div class="spacer"></div>
  </div>
</header>
`;

exports[`AppHeader should be exist right item 1`] = `
<header class="v-sheet v-sheet--tile theme--light v-toolbar v-toolbar--dense v-toolbar--flat primary" style="height: 48px;">
  <div class="v-toolbar__content" style="height: 48px;">
    <div title="vs-minesweeper" class="v-toolbar__title">
      VSマインスイーパー
    </div>
    <div class="spacer"></div> <button type="button" class="v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--default" title="information"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-information theme--light"></i></span></button>
  </div>
</header>
`;

exports[`AppHeader should be have title by default 1`] = `
<header class="v-sheet v-sheet--tile theme--light v-toolbar v-toolbar--dense v-toolbar--flat primary" style="height: 48px;">
  <div class="v-toolbar__content" style="height: 48px;">
    <div title="vs-minesweeper" class="v-toolbar__title">
      VSマインスイーパー
    </div>
    <div class="spacer"></div>
  </div>
</header>
`;
